### PR TITLE
Handle paritial nonblock write

### PR DESCRIPTION
**Pull request by ganmacs**
closes https://github.com/fluent/fluent-logger-ruby/issues/78

### CODE REVIEW

File: lib/fluent/logger/fluent_logger.rb [hunk 1]
```diff
@@ -230,8 +230,12 @@ def write(msg)
           end
 
           begin
-            send_data(@pending)
-            @pending = nil
+            len = send_data(@pending)
+            if len != pending_bytesize
+              @pending = @pending.slice(len, pending_bytesize)
```
Comment (repeatedly) [inline lib/fluent/logger/fluent_logger.rb]: If slice the pending data, next `send_data` sends broken data, right?

Comment (ganmacs) [reply lib/fluent/logger/fluent_logger.rb]: Good point… this implementation doesn't work 😢 
It is necessary to send msgpacked data all at once in one call of Fluent::Logger#post.

Comment (ganmacs) [reply lib/fluent/logger/fluent_logger.rb]: fixed. https://github.com/fluent/fluent-logger-ruby/pull/80/commits/8d53f600c96062faf503068d4f533775b1f0601a

File: lib/fluent/logger/fluent_logger.rb [hunk 2]
```diff
@@ -272,7 +276,19 @@ def send_data(data)
...
         #end
-        true
+      end
+
+      def send_data_nonblock(data)
+        written = @con.write_nonblock(data)
+        remaining = data.bytesize - written
+
+        while remaining > 0
+          len = @con.write_nonblock(data.byteslice(written, remaining))
```
Comment (repeatedly) [inline lib/fluent/logger/fluent_logger.rb]: should check sliced data is valid msgpack or not?

Comment (ganmacs) [reply lib/fluent/logger/fluent_logger.rb]: You mean that the value `data.byteslice(written, remaining)` returns should be checked if it is valid msgpack or not?  Are there any problems if this value is not valid msgpack?

Comment (repeatedly) [reply lib/fluent/logger/fluent_logger.rb]: Ah, I forgot this write uses same socket.

**Comment by repeatedly** [comment]
Thanks!

**Pull request merged**